### PR TITLE
#9 integration tests

### PR DESCRIPTION
**Pull request by danilakliashchou**
/device_backup/{deviceId} endpoint GET/POST methods integration test

### CODE REVIEW

File: tests/integration/test_save_get.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from client import ApiClient
+import uuid
+import pytest
+
+test_uuid = uuid.uuid1()
+
+
+def test_save(api_client: ApiClient):
```
Comment (hleb-albau) [inline tests/integration/test_save_get.py:8]: Давай немного переделаем тесты
Один из тестов будет полноценный позитивный сценарий
```
#  Test scenario:
#  1. Check, that backup for 'device_id' doesn't exists
#  2. Save 'device_id' **test_blob1** and verify server returns 200
#  3. Check metadata exists for 'device_id' with timestamp near save point (tbd when head endpoint will be done)
#  4. Get 'device_id' backup, check received blob equal **test_blob1**
#  5. Update 'device_id' backup with another **test_blob2**
#  6. Check a new metadata exists for 'device_id' with timestamp near new save point (tbd when head endpoint will be done)
#  7. Get 'device_id' backup, check received blob equal **test_blob2**
```

File: tests/integration/client.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    base_url: str
+
+    def __init__(self, base_url):
+        self.base_url = base_url
+
+    def test_get_backup(self, device_uuid: str):
+        path = self.base_url + f"/device_backup/{device_uuid}"
+        return requests.get(url=path, verify=False)
+
+    def test_save_backup(self, device_uuid: str, body: str):
```
Comment (hleb-albau) [inline tests/integration/client.py:24]: 1. тут можно без test_ приставок, ниже у тебя так и используется(как оно работает тогда?)
2. боди должно быть байтами, а не строкой

File: tests/integration/client.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+import pytest
+import requests
+
+from src.consts.env_url import DEVICE_BACKUP_SERVICE_BASE_URL
+
+
+@pytest.fixture(scope="session")
+def api_client():
+    return ApiClient(os.environ[DEVICE_BACKUP_SERVICE_BASE_URL])
```
Comment (hleb-albau) [inline tests/integration/client.py:11]: вместо env переменной можно использовать **pytest-base-url** пакет

File: tests/integration/test_save_get.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    valid_data = ""
+    valid_data = [valid_data + str(i) for i in range(20000)]
```
Comment (hleb-albau) [inline tests/integration/test_save_get.py:10]: Большинство ошибок в программировании находятся в так называемых Corner-case, Edge case, Boundary case, соответственно их нужно в первую очередь и включать в тесты. Обычно это не позитивный флоу, когда твои сервисом пользуются правильно, а отрицательный, когда его используют неправильно.

Например для метода save можно задать следующие вопросы:
- а что если передать пустое боди?
- а что если мы не передадим device_id
- а что если мы передадим device_id не uuid

И так далее для всех методов) Может еще случаи продумаешь

File: tests/integration/test_save_get.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    assert response.status_code == 200
+
+
+def test_get(api_client: ApiClient):
+    response = api_client.get_backup(device_uuid=test_uuid)
+    assert response.status_code == 200
+
+
+def test_save_invalid_data(api_client: ApiClient):
+    invalid_data = "testInvalidData"
```
Comment (hleb-albau) [inline tests/integration/test_save_get.py:21]: не понятно, почему это невалидные данные из теста
лучше использовать более понятные нейминги(то, что я до этого в прошлом пре писал, более конкретные и мнее абстрактные) для теста и самим данных
например:
вместо `invalid_data` -> `less_than_10kb_data`
test_save_invalid_data -> `test_save_less_than_10kb_data_backup`

В целом писать нужно так, чтобы человек, который не зная контекста, мог понять максимально смысла происходящего

File: tests/integration/test_save_get.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    response = api_client.post_backup(device_uuid=uuid.uuid1(), body=invalid_data)
+    assert response.status_code == 400
+
+
+def test_get_nonexistent(api_client: ApiClient):
+    response = api_client.get_backup(device_uuid="nonexistentBackup")
```
Comment (hleb-albau) [inline tests/integration/test_save_get.py:27]: для несуществующего backup вроде должно быть 404, а не 400

**Review by hleb-albau — CHANGES_REQUESTED** [review]

File: tests/integration/test_save_get.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    check_id_exists = api_client.get_backup(device_uuid=str(test_uuid))
+    assert check_id_exists.status_code == 404
+
+    # 2
+    save_backup = api_client.save_backup(device_uuid=str(test_uuid), body=bytes(valid_data, encoding="utf-8"))
+    assert save_backup.status_code == 200
+
+    # 4
+    get_device_backup = api_client.get_backup(device_uuid=str(test_uuid))
+    assert get_device_backup.text == str(bytes(valid_data, encoding="utf-8"))
```
Comment (hleb-albau) [inline tests/integration/test_save_get.py:34]: вообще можно сразу bytes делать и байты между собой сравнивать

File: tests/integration/test_save_get.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+#  4. Get 'device_id' backup, check received blob equal **test_blob1**
+#  5. Update 'device_id' backup with another **test_blob2**
+#  6. Check a new metadata exists for 'device_id' with timestamp near new save point (tbd when head endpoint will be done)
+#  7. Get 'device_id' backup, check received blob equal **test_blob2**
+
+
+def test_positive_scenario(api_client: ApiClient):
+
+    # 1
+    test_uuid = uuid.uuid1()
```
Comment (hleb-albau) [inline tests/integration/test_save_get.py:24]: еще раз генерим test_uuid?
в целом данные для теста лучше чтобы или все в нчале определялись, или перед тестом (но в одном месте)

File: tests/integration/test_save_get.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    assert response.status_code == 404
+
+
+def test_empty_body(api_client: ApiClient):
+    response = api_client.save_backup(device_uuid=str(test_uuid), body=bytes("", encoding="utf-8"))
+    assert response.status_code == 400
+
+
+def test_not_uuid(api_client: ApiClient):
+    get_response = api_client.get_backup(device_uuid="notUuidId")
```
Comment (hleb-albau) [inline tests/integration/test_save_get.py:65]: лучше всегда проверять status_code

**Review by hleb-albau — CHANGES_REQUESTED** [review]

**Review by hleb-albau — APPROVED** [review]
